twoliter/embedded/img2img

```diff
@@ -1,3 +1,318 @@
+[[ -s "${GRUB_CONFIG}" ]]
+
+# Replace the dm-verity root with the new verity
+sed -i \
+  "s:^set dm_verity_root=.*:set dm_verity_root=\"${DM_VERITY_ROOT[*]}\":g" \
+  "${GRUB_CONFIG}"
+
+# Replace grub.cfg on the boot image.
+cat <<EOF | debugfs -w -f - "${BOOT_IMAGE}"
+rm /grub/grub.cfg
+write ${GRUB_CONFIG} /grub/grub.cfg
+ea_set /grub/grub.cfg security.selinux system_u:object_r:os_t:s0
+EOF
+
+# Sign the grub.cfg and replace the signature on the boot image, if needed.
+if [[ "${UEFI_SECURE_BOOT}" == "yes" ]]; then
+  sign_grubcfg "${GRUB_CONFIG}"
+  cat <<EOF | debugfs -w -f - "${BOOT_IMAGE}"
+rm /grub/grub.cfg.sig
+write ${GRUB_CONFIG}.sig /grub/grub.cfg.sig
+ea_set /grub/grub.cfg.sig security.selinux system_u:object_r:os_t:s0
+EOF
+fi
+
+# Write the boot image back to the OS image.
+check_image_size "${BOOT_IMAGE}" "${partsize["BOOT-A"]}"
+dd if="${BOOT_IMAGE}" of="${OS_IMAGE}" \
+  conv=notrunc bs=1M seek="${partoff["BOOT-A"]}"
+
+###############################################################################
+# Section 7: generate final artifacts and copy to output dir
+
+# Panic even for warnings, such as when the main and backup tables differ.
+if OS_IMAGE_VALIDATION=$(sgdisk -v "${OS_IMAGE}"); then
+  [[ "${OS_IMAGE_VALIDATION}" =~ "Warning!" ]] && exit 1
+else
+  exit 1
+fi
+
+# Re-compress the OS image and generate OS/DATA symlinks.
+if [[ "${OUTPUT_FMT}" == "raw" ]]; then
+  compress_image "img.lz4" "os_image" "${OUTPUT_DIR}"
+  symlink_image "img.lz4" "os_image" "${OUTPUT_DIR}"
+  if [[ -s "${DATA_IMAGE}" ]]; then
+    symlink_image "img.lz4" "data_image" "${OUTPUT_DIR}"
+  fi
+elif [[ "${OUTPUT_FMT}" == "qcow2" ]]; then
+  compress_image "qcow2" "os_image" "${OUTPUT_DIR}"
+  symlink_image "qcow2" "os_image" "${OUTPUT_DIR}"
+  if [[ -s "${DATA_IMAGE}" ]]; then
+    symlink_image "qcow2" "data_image" "${OUTPUT_DIR}"
+  fi
+elif [[ "${OUTPUT_FMT}" == "vmdk" ]]; then
+  compress_image "vmdk" "os_image" "${OUTPUT_DIR}"
+  symlink_image "vmdk" "os_image" "${OUTPUT_DIR}"
+  if [[ -s "${DATA_IMAGE}" ]]; then
+    symlink_image "vmdk" "data_image" "${OUTPUT_DIR}"
+  fi
+fi
+
+# Create the OVA if needed.
+if [[ "${OUTPUT_FMT}" == "vmdk" ]]; then
+  generate_ova \
+    "${OS_IMAGE_NAME}.vmdk" \
+    "${DATA_IMAGE_NAME}.vmdk" \
+    "${OS_IMAGE_PUBLISH_SIZE_GIB}" \
+    "${DATA_IMAGE_PUBLISH_SIZE_GIB}" \
+    "${OVF_TEMPLATE}" \
+    "${UEFI_SECURE_BOOT}" \
+    "${OUTPUT_DIR}"
+  symlink_image "ova" "os_image" "${OUTPUT_DIR}"
+fi
+
+# Compress and symlink the rest.
+compress_image "ext4.lz4" "boot_image" "${OUTPUT_DIR}"
+compress_image "verity.lz4" "verity_image" "${OUTPUT_DIR}"
+compress_image "ext4.lz4" "root_image" "${OUTPUT_DIR}"
+
+symlink_image "ext4.lz4" "boot_image" "${OUTPUT_DIR}"
+symlink_image "verity.lz4" "verity_image" "${OUTPUT_DIR}"
+symlink_image "ext4.lz4" "root_image" "${OUTPUT_DIR}"
+
+popd >/dev/null
+
+# Ensure proper ownership of the final artifacts.
+find "${OUTPUT_DIR}" -type f -print -exec chown 1000:1000 {} \;
```

Comment:
Might also be worth cleaning up the temp files, as `rpm2img` does.

Reply:
Towards the top of the script there is a trap to clean-up **WORKDIR**, which is where all the temp files are consolidated. Would you prefer it just be at the end like in `rpm2img`?

Reply:
No, the trap is better, I just missed the implicit cleanup.